rust/src/syscore.rs

```diff
@@ -0,0 +1,82 @@
+pub use self::ffi::*;
+use std::ptr::NonNull;
+
+/// Opaque type for C interop: RpmOstreeOrigin.
+pub enum RpmOstreeOrigin {}
+
+extern "C" {
+    fn rpmostree_origin_get_live_state(
+        origin: *const RpmOstreeOrigin,
+        out_inprogress: *mut *mut libc::c_char,
+        out_livereplaced: *mut *mut libc::c_char,
+    );
+}
+
+/// Reference to an `RpmOstreeOrigin`.
+#[derive(Debug)]
+pub struct OriginRef {
+    origin: NonNull<RpmOstreeOrigin>,
+}
+
+impl OriginRef {
+    /// Build a reference object from a C pointer.
+    fn from_ffi_ptr(oref: *mut RpmOstreeOrigin) -> Self {
+        Self {
+            origin: NonNull::new(oref).expect("NULL RpmOstreeOrigin"),
+        }
+    }
+
+    /// Get `livefs` details for this deployment origin.
+    pub fn get_live_state<'o>(&'o self) -> OriginLiveState<'o> {
+        use crate::ffiutil::ffi_new_nullable_string;
+
+        let mut out_inprogress: *mut libc::c_char = std::ptr::null_mut();
+        let mut out_livereplaced: *mut libc::c_char = std::ptr::null_mut();
+        unsafe {
+            rpmostree_origin_get_live_state(
+                self.origin.as_ptr(),
+                &mut out_inprogress,
+                &mut out_livereplaced,
+            )
+        };
+        let in_progress = ffi_new_nullable_string(out_inprogress);
```

Comment:
I think this might exist today via `glib::translate`; try something like this:
`let in_progress: Option<String> = from_glib_full(out_inprogress);`?

See http://gtk-rs.org/docs/src/glib/translate.rs.html#1347-1356

Reply:
(And I suspect some of the other `ffi_view_` things should just be `from_glib_none` but at the time I didn't know about everything in GLib Rust bindings)

Reply:
Lovely, if that works it will also spare the `g_free` below. I guess I can also drop the custom `ffi_` helper I've introduced.